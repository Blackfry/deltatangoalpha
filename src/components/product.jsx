import React from 'react';
import Panel from 'react-bootstrap/lib/Panel';
import Grid from 'react-bootstrap/lib/Grid';
import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';
import { Link } from 'react-router';
import classNames from 'classnames';


export const Product = props => {
    const {id, name, image, code, view, noLink} = props;

    // Product is reused in list and details views ando is returned in a larger format for detail.
    let imgContClass = classNames("Product-image-container", view);
    let codeTextClass = classNames("Product-code-text", ((view === 'detail' && code !== '') ? 'show' : 'hidden'));

    return (
        <div className="animated fadeIn">
            <Link to={(noLink ? '/addproduct/' : '/products/'+id+'/')}>
                <Panel>
                    <Grid style={{
                        width: '100%'
                    }}>
                        <Row>
                            <Col>
                                <div className={imgContClass}>
                                    <div style={{
                                        background: 'url(' + image + ')',
                                        height: '100%',
                                        backgroundSize: 'cover',
                                        backgroundPosition: 'center',
                                    }}></div>
                                </div>
                            </Col>
                        </Row>
                        <Row>
                            <Col>
                                <div>
                                    <h4>{name}</h4>
                                    <h6 className={codeTextClass}>{code}</h6>
                                </div>
                            </Col>
                        </Row>
                    </Grid>
                </Panel>
            </Link>
        </div>
    )
};


Product.propTypes = {
    id: React.PropTypes.number.isRequired,
    name: React.PropTypes.string.isRequired,
    image: React.PropTypes.string.isRequired,
    code: React.PropTypes.string
};

Product.defaultProps = {
    id: 1,
    name: "There was a problem showing your product, try refreshing the page.",
    image: "https://img.buzzfeed.com/buzzfeed-static/static/2015-01/26/21/enhanced/webdr09/enhanced-1014-1422325554-19.jpg",
    code: "",
    view: "detail"
};



