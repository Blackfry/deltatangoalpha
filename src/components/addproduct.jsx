import React, {Component} from 'react';
import {connect} from 'react-redux';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {Field, reduxForm} from 'redux-form'
import {RaisedButton, TextField} from 'material-ui'
import {addNewProductInitiateActionCreator, loadProductsInitiateActionCreator} from '../redux/actions';
import injectTapEventPlugin from 'react-tap-event-plugin';
import {Panel} from 'react-bootstrap/lib/Panel';
import {Product} from './product'

// https://github.com/callemall/material-ui/issues/4670
injectTapEventPlugin();

// const required = value => value == null ? 'Required' : undefined;


const renderTextField = ({input, context, label, meta: {touched, error}, ...custom})=> (
    <TextField
        hintText={label}
        floatingLabelText={label}
        errorText={touched && error}
        {...input}
        {...custom}
    />
);


let AddProductForm = props => {
    const {handleSubmit, pristine, submitting} = props;

    return (
        <MuiThemeProvider>
            <div style={{padding: '30px'}}>
                <form onSubmit={handleSubmit}>
                    <div>
                        <Field
                            name="name"
                            component={renderTextField}
                            label="Name"
                            hintText="min 10 characters required"/>
                    </div>
                    <div>
                        <Field
                            name="image"
                            component={renderTextField}
                            label="Image"
                            hintText="url must start with 'http'"/>
                    </div>
                    <div>
                        <Field
                            name="promocode"
                            component={renderTextField}
                            label="Promo Code"
                            hintText="Optional"/>
                    </div>
                    <div>
                        <RaisedButton
                            label="Create Product"
                            primary={true}
                            style={{margin: '12px'}}
                            type="submit"
                            disabled={pristine || submitting}
                        />
                    </div>
                </form>
            </div>
        </MuiThemeProvider>
    )
};


const validate = values => {
    const errors = {};
    const requiredFields = ["name", "image",];
    requiredFields.forEach(field => {
        if (!values[field]) {
            errors[field] = 'Required'
        }
    });

    if (values.name  && values.name.length < 10) {
        errors.name = 'Product name must be minimum 10 characters'
    }
    if (values.image && values.image.slice(0, 4) !== 'http') {
      errors.image = 'Image url must start with http'
    }
    return errors
};


AddProductForm = reduxForm({
    form: 'addProductForm',
    validate
})(AddProductForm);


const mapStateToProps = (state) => ({
    PRODUCT: state.PRODUCT,
    form: state.form
});

const mapDispatchToProps = (dispatch) => ({
    dispatch,
});


export class AddProductFormContainer extends Component {

    componentWillMount() {
        if (!this.props.PRODUCT.productListLoaded) {
            this.props.dispatch(loadProductsInitiateActionCreator())
        }
    }

    asyncFormSubmitHandler = productObj => {
        this.props.dispatch(addNewProductInitiateActionCreator(productObj));

    };

    render() {
        if(this.props.PRODUCT.addNewProductSuccess) {
            return (
                <div>
                    <div className="animated slideInDown">
                        <div className="Detail-container">
                            <div className="Product-detail">
                                <Product
                                    {...this.props.PRODUCT.newProduct}
                                    view="detail"
                                />
                            </div>
                        </div>
                    </div>
                    <div className="Product-form-container">
                        <div className="Product-form">
                            <Panel>
                                <div>
                                    <AddProductForm
                                        dispatch={this.props.dispatch}
                                        onSubmit={ this.asyncFormSubmitHandler }
                                    />
                                </div>
                            </Panel>
                        </div>
                    </div>
                </div>
            )
        }

        if(!this.props.PRODUCT.addNewProductSuccess) {
            return (
                <div className="Product-form-container">
                    <div className="Product-form">
                        <Panel>
                            <div>
                                <AddProductForm
                                    dispatch={this.props.dispatch}
                                    onSubmit={ this.asyncFormSubmitHandler }
                                />
                            </div>
                        </Panel>
                    </div>
                </div>
            )
        }
    }
}


export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AddProductFormContainer);