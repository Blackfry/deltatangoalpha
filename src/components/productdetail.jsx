import React, { Component } from 'react';
import { connect } from 'react-redux';
import { loadProductsInitiateActionCreator } from '../redux/actions';
import { Product } from './product'


const mapStateToProps = (state) => ({
    PRODUCT: state.PRODUCT
});

const mapDispatchToProps = (dispatch) => ({
    dispatch,
});


export class ProductDetailContainer extends Component {

    componentWillMount() {
        // make sure the product list is loaded
        let productId = this.props.routeParams.id;
        if (!this.props.PRODUCT.productListLoaded && typeof productId !== 'undefined') {
            this.props.dispatch(loadProductsInitiateActionCreator())
        }
    }

    render() {
        if(!this.props.PRODUCT.productListLoaded) {
            return (
                <div>
                    ...loading
                </div>
            )
        }

        // search the list of products for one with the unique id we are looking for.
        // errors where no product is found are handled by defaultProps in the Product component
        let product = this.props.PRODUCT.products.filter(x => x.id.toString() === this.props.routeParams.id)[0];

        return (
            <div className="Detail-container">
                <div className="Product-detail">
                    <Product {...product} view="detail"/>
                </div>
            </div>
        )
    }
}


export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ProductDetailContainer);
