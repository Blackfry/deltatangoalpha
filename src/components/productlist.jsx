import React, { Component } from 'react';
import { connect } from 'react-redux';
import { loadProductsInitiateActionCreator } from '../redux/actions';
import { Product } from './product';
import Grid from 'react-bootstrap/lib/Grid';
import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';


const mapStateToProps = (state) => ({
    PRODUCT: state.PRODUCT
});

const mapDispatchToProps = (dispatch) => ({
    dispatch,
});


const composeProductList = props => {
    const { products } = props;

    return products.map((product, i) => {
        return (
            <div key={i} className="animated fadeIn">
                 <Col xs={12} sm={6} md={4} lg={3}>
                    <Product {...product} view="list"/>
                 </Col>
            </div>
        )
    })
};


export class ProductListContainer extends Component {

    componentWillMount() {
        if (!this.props.PRODUCT.productListLoaded) {
            this.props.dispatch(loadProductsInitiateActionCreator())
        }
    }


    render() {
        if(!this.props.PRODUCT.productListLoaded) {
            return (
                <div>
                    ...loading
                </div>
            )
        }

        let products = this.props.PRODUCT;
        return (
            <div className="Detail-container">
                <Grid>
                    <Row>
                        <div className="Product-list">
                            {composeProductList(products)}
                        </div>
                    </Row>
                </Grid>
            </div>
        )
    }
}



export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ProductListContainer);
