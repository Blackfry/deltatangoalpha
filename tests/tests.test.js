import React from 'react';
import {
    mount, shallow
} from 'enzyme';
import renderer from 'react-test-renderer';
import { Product } from '../src/components/product';

const productData =
{
    "products": [
        {
            "id": 1,
            "name": "product 1",
            "image": "http://lorempixel.com/300/250/city/image%201",
            "code": ""
        },
        {
            "id": 2,
            "name": "product 2",
            "image": "http://lorempixel.com/300/250/city/image%202",
            "code": "promo1"
        },
        {
            "id": 3,
            "name": "product 3",
            "image": "http://lorempixel.com/300/250/city/image%203",
            "code": ""
        },
        {
            "id": 4,
            "name": "product 4",
            "image": "http://lorempixel.com/300/250/city/image%204",
            "code": "promo1"
        },
    ]
}


describe('(Component) Product', () => {
    it('test rendering component produces something', () => {
        const product = productData.products[0];
        const wrapper = shallow(
            <Product {...product}/>
        );
        expect(wrapper).toHaveLength(1);
    });
});


test('Testing rendering of Product component', () => {
    const product = productData.products[0];
    const component = renderer.create(
        <Product {...product}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});