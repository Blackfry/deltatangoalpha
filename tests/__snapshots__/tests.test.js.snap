exports[`test Testing rendering of Product component 1`] = `
<div
  className="animated slideInRight">
  <a
    onClick={[Function]}
    style={Object {}}>
    <div
      className="panel panel-default"
      id={undefined}>
      <div
        className="panel-body">
        <div
          className="container"
          style={
            Object {
              "width": "100%",
            }
          }>
          <div
            className="row">
            <div
              className="">
              <div
                className="Product-image-container detail">
                <div
                  style={
                    Object {
                      "background": "url(http://lorempixel.com/300/250/city/image%201)",
                      "backgroundPosition": "center",
                      "backgroundSize": "cover",
                      "height": "100%",
                    }
                  } />
              </div>
            </div>
          </div>
          <div
            className="row">
            <div
              className="">
              <div
                style={
                  Object {
                    "fontSize": "20px",
                    "overflow": "hidden",
                    "textOverflow": "ellipsis",
                  }
                }>
                <h4>
                  product 1
                </h4>
                <h6
                  className="Product-code-text hidden">
                  
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </a>
</div>
`;
